docs/reactive-patterns.qmd

```diff
@@ -0,0 +1,232 @@
+
+ui.input_slider("n", "N", min=1, max=100, value=1)
+ui.input_action_button("compute", "Compute!")
+
+@render.text
+async def result():
+    input.compute()        # Take a dependency on the button
+    await asyncio.sleep(2) # Simulate a long computation
+    with reactive.isolate():
+        # Read input.n() without taking a dependency on it
+        return f"Result: {input.n()}"
+```
+
+:::
+
+
+## Requiring input {#req}
+
+Sometimes you want to require an input before executing a reactive function and/or displaying output. The `req()` is designed for this purpose. It effectively stops execution until the statement provided is truthy.
+
+For example, the app below allows a user to upload a csv file, which is then used to render a table.
+Notice how the reactive calculation, `df`, uses `req()` to stop execution until the user has uploaded a file.
+
+
+::: {data-shinylive="https://shinylive.io/py/editor/#code=NobwRAdghgtgpmAXGKAHVA6VBPMAaMAYwHsIAXOcpMASxlWICcyACVKCAEygGcXe2nADoQAZo2IwWPABY0I2FnQbMWjOFEJkaANzh41lTnEYH1ARwMBXGiPGTpchRjgAPVOp59lTVvNRWZCIiNhj+gQD6ojQANnAAFEJg0XFJBkkAqqgxxFCc-CyEPDosKXBp-ISEcKhkALxJGEU6SQCUwRAAAuqa2npNUDGEIsaiLJyi8a2IIixzpSx1hubx4WQYZVPtEPOGZFaMO6icGD2cEc3xosAADAC6wADk3GRoUGQyj3fbIt1GJhgKK4glw4GMgWQpjMdvN1PtDqUkgBBFggCZTDCyNBwW53AC+rjRk1amJk2OAAEZ8eN3lBxLBymAOn9QYwMC8oFFGAyRmCWK8AEZxKGzWFweE7dSsjAAEVpAHFGDROPF0a1WmA8XcgA"}
+
+```{python}
+#| eval: false
+import pandas as pd
+from shiny import reactive, req
+from shiny.express import input, renderm ui
+
+ui.input_file("file", "Upload a csv file", accept=".csv")
+
+@reactive.calc
+def df():
+    # req() stops execution until input.file() is truthy
+    f = req(input.file())
+    return pd.read_csv(f[0]['datapath'])
+
+@render.data_frame
+def table():
+    # Output won't render until input.file() is truthy
+    return render.DataGrid(df())
+```
+:::
+
+![A gif of uploading a csv file and seeing the text and table outputs update](assets/file-upload.mp4){class="img-fluid shadow rounded" style="margin-top: -1.5rem;"}
+
+
+## Scheduled invalidation {#invalidate-later}
+
+Sometimes its useful invalidate a reactive function on a schedule.
+For example, to implement a clock that updates every second, you can use `reactive.invalidate_later(1)`:
+
+```{shinylive-python}
+#| standalone: true
+#| components: [editor, viewer]
+#| layout: vertical
+#| viewerHeight: 50
+from datetime import datetime
+from shiny import reactive
+from shiny.express import render
+
+# Get the current time every second
+@reactive.calc
+def cur_time():
+    reactive.invalidate_later(1)
+    return datetime.now().strftime('%H:%M:%S')
+
+@render.ui
+def clock():
+    return f"Current time: {cur_time()}"
+```
+
+
+## Reactive file reading {#file}
+
+Sometimes its useful invalidate a reactive function when a file changes.
+For example, lets extend the example from above to write the current time to a file every second, and then read and display the contents of that file:
```

Comment:
This isn't that compelling of an example to me, it might be better to say something like "Imagine an app that depends on a csv file which is populated by an Airflow pipeline, you want the app to always show the most up-to-data data without needing to restart the user's session". the plot will update whenever the file changes on disk. You can't really show the interaction in the same way but I think it's better to have an example which solves a real user problem.